spec/java/util/StringBuilder.main.lsl

```diff
@@ -0,0 +1,1022 @@
+    }
+
+
+    @Phantom proc _replace_loop (i: int, strIndex: int, start: int, end: int, newStr: array<char>, str: String, strLength: int): void
+    {
+        if (i < start)
+        {
+            newStr[i] = action CALL_METHOD(this.storage, "charAt", [i]);
+        }
+        else if (i < start + strLength)
+        {
+            newStr[i] = action CALL_METHOD(str, "charAt", [strIndex]);
+            strIndex += 1;
+        }
+        else
+        {
+            newStr[i] = action CALL_METHOD(this.storage, "charAt", [end]);
+            end += 1;
+        }
+    }
+
+
+    fun *.reverse (@target self: StringBuilder): StringBuilder
+    {
+        var n: int = this.length - 1;
+        var i: int = (n-1) / 2;
+        var k: int = -1;
+        val newStr: array<char> = action ARRAY_NEW("char", this.length);
+        action LOOP_FOR(i, 0, -1, -1, _reverse_loop(i, k, n, newStr));
+
+        this.storage = action OBJECT_TO_STRING(newStr);
+        result = self;
+    }
+
+
+    @Phantom proc _reverse_loop (i: int, k: int, n: int, newStr: array<char>): void
+    {
+        k = n - i;
+        newStr[k] = action CALL_METHOD(this.storage, "charAt", [i]);
+        newStr[i] = action CALL_METHOD(this.storage, "charAt", [k]);
+    }
+
+
+    fun *.toString (@target self: StringBuilder): String
+    {
+        action NOT_IMPLEMENTED("no final decision");
+    }
+
+
+    // within java.lang.AbstractStringBuilder
+    fun *.capacity (@target self: StringBuilder): int
+    {
+        // This is right realization ?
+        // original: return value.length >> coder;
+        // we will have another result, because we don't know "coder" value;
+        result = this.length;
+    }
+
+
+    // within java.lang.AbstractStringBuilder
+    fun *.ensureCapacity (@target self: StringBuilder, minimumCapacity: int): void
+    {
+
+    }
+
+
+    // within java.lang.AbstractStringBuilder
+    fun *.length (@target self: StringBuilder): int
+    {
+        result = this.length;
+    }
+
+
+    // within java.lang.AbstractStringBuilder
+    fun *.charAt (@target self: StringBuilder, index: int): char
+    {
+        _checkIndex(index);
+        result = action CALL_METHOD(this.storage, "charAt", [index]);
+    }
+
+
+    // within java.lang.AbstractStringBuilder
+    fun *.setLength (@target self: StringBuilder, newLength: int): void
+    {
+        if (newLength < 0)
+        {
+            //val message: String = "String index out of range: " + action OBJECT_TO_STRING(newLength);
+            action THROW_NEW("java.lang.StringIndexOutOfBoundsException", []);
+        }
+        var i: int = 0;
+        val newStr: array<char> = action ARRAY_NEW("char", newLength);
+        action LOOP_FOR(i, 0, newLength, +1, _setNewLength_loop(i, newStr));
+
+        this.storage = action OBJECT_TO_STRING(newStr);
+        this.length = newLength;
+    }
+
+
+    @Phantom proc _setNewLength_loop (i: int, newStr: array<char>): void
+    {
+        if (i < this.length)
```

Comment:
Loop bound calculations should be done before it even starts.
`else` branch is not needed: `this.storage` IS a `String` thus we do not care about trailing `null` characters.